### PR TITLE
Spot instance support

### PR DESCRIPTION
**Pull request by dalehamel**
WIP

This is an attempt to close https://github.com/hashicorp/terraform/issues/1339.

I'd like some feedback on the approach I've taken here before I smooth off the rough edges

Initially I thought it would be a good idea to implement spot instances as a separate resources, but since spot instances share almost all of the same attributes as normal on demand instances, this isn't very DRY.

So, the approach I've taken here is to instead add spot related attributes to the aws_instance resource. If 'spot_price' is specified, it will attempt to launch a spot instance. It will fail if the spot instance provisioning fails **unless** the 'spot_fallback' option is specified, in which case it will just execute the normal on-demand provisioning path.

The code changes here do a few things:

* Add the necessary resource data for spot instances (spot_price, spot_fallback). I plan on also adding a 'spot_persist' flag, so that the spot instance can be automatically relaunched if the spot instance is preempted by amazon.
* The provisioning of the on demand instance is now conditional on the request not being for a spot instance **or** a spot request failing, and the spot_fallback option being set.
* Spot requests will take advantage of all of the attribute parsing that is done in the runOpts for on demand instances, to keep things DRY.

There are several things that I recognize need to be fixed before this can be merged:

* [x] delete should delete the spot request
* [x] all other FIXMEs should be addressed
* [x] 2 spaces should be converted to tabs to be consistent with the rest of the code

@ketzacoatl @mitchellh @catsby  for review

### CODE REVIEW

**Comment by dalehamel** [comment]
FYI this is also my first time writing any Go code, so please don't be afraid to nitpick for styles / Go idioms I don't know.

**Comment by cbednarski** [comment]
@dalehamel Can you run `go fmt` on your checkins? Go uses hard tabs for indentation. The formatting tool will fix it for you. Reading through the rest of it atm.

**Comment by dalehamel** [comment]
thanks @cbednarski, thanks handy - done.

File: builtin/providers/aws/resource_aws_instance.go [hunk 1]
```diff
@@ -514,44 +530,130 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
...
+
+		request_id := *spotResp.SpotInstanceRequests[0].SpotInstanceRequestID
+		d.Set("spot_request_id", request_id)
+
+		spotStateConf := &resource.StateChangeConf{
+			// http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html
+			Pending:    []string{"start", "pending-evaluation", "pending-fulfillment"},
+			Target:     "fulfilled",
+			Refresh:    SpotInstanceStateRefreshFunc(conn, request_id),
+			Timeout:    10 * time.Minute,
```
Comment (cbednarski) [inline builtin/providers/aws/resource_aws_instance.go]: It looks like the script may wait 10 minutes for a bid to complete. You can't run `terraform apply` in parallel so In the mean time, you will be unable to apply other changes to your infrastructure. How likely is it that you'll wait the full 10 minutes? Can this be shorter (e.g. 10-30s)? Or configurable?

File: builtin/providers/aws/resource_aws_instance.go [hunk 2]
```diff
@@ -514,44 +530,130 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
...
+
+		log.Printf("[DEBUG] waiting for spot bid to resolve... this may take several minutes.")
+		spotRequestRaw, err := spotStateConf.WaitForState()
+		spotRequest := spotRequestRaw.(*ec2.SpotInstanceRequest)
+
+		fallback := d.Get("spot_fallback").(bool)
+
+		if err != nil && !fallback {
+			return fmt.Errorf("Error while waiting for spot request (%s) to resolve: %s", request_id, err)
+		} else if fallback {
```
Comment (cbednarski) [inline builtin/providers/aws/resource_aws_instance.go]: Here you fall back to a standard instance if the spot bid fails, but spot instances may succeed now and be pre-empted later. In that scenario the on-demand instance will not spin up to replace it unless / until you run `terraform apply` again.

I wonder if it would make more sense to have spot be a separate type of resource. Any thoughts about this?

File: builtin/providers/aws/resource_aws_instance.go (hunk 1 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: This all comes down to how fast amazon resolves your bid. Typically it's about 5 minutes or less, but I've seen it take as long as half an hour in rare cases.

Making it configurable wouldn't hurt - to be honest though this is just copy pasta. If someone decides they want a spot instance, they should be aware that it will take some time for their bid to resolve

File: builtin/providers/aws/resource_aws_instance.go (hunk 2 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: > Here you fall back to a standard instance if the spot bid fails, but spot instances may succeed now and be pre-empted later.

I hadn't considered that edge case when writing this, thanks. This will only be the case if the request actually times out. The logic i had considered when writing was that the spot instance fails because the price does not resolve (bid too low) or amazon otherwise cannot provision the capacity. The intention of 'spot fallback' is to give a pricier on demand instance if the spot request cannot be fulfilled, aka 'try to get the cheap instance, but settle for full price'.

> I wonder if it would make more sense to have spot be a separate type of resource. Any thoughts about this?

I'm ambivalent on this too. On the one hand, having it as a separate resource feels cleaner. On the other hand, it's not very DRY because spot instances and on demand instances are almost identical - the only difference is how they are launched. Spot instances have the extra 'bid resolution' phase, but aside from that most of the parameters are the same. Making them a separate resource would basically entail copying this file and maintaining all this logic in two places.

Are spot instances really different than normal AWS instances? No, they aren't - they're identical instances. The only difference is the provisioning strategy. Spot instances take longer to provision, and provisioning is less certain. The upside though, is that you can typically save a **substantial** amount of money by using them, and they fit a lot of use cases (particularly for ephemeral compute resources) very nicely. So, if you're willing to wait, it can be worth it.

File: builtin/providers/aws/resource_aws_instance.go (hunk 1 above)
Comment (cbednarski) [inline builtin/providers/aws/resource_aws_instance.go]: Do I actually need to wait for the bid to resolve? Or can I just post it and move on, and assume my instances may spin up later? When I read the docs on spot I got the impression that you can post a bid and walk away and AWS will boot the instances for you later if the price drops. Maybe that feature is optional?

File: builtin/providers/aws/resource_aws_instance.go (hunk 2 above)
Comment (cbednarski) [inline builtin/providers/aws/resource_aws_instance.go]: > Spot instances take longer to provision, and provisioning is less certain. The upside though, is that you can typically save a substantial amount of money by using them, and they fit a lot of use cases (particularly for ephemeral compute resources) very nicely.

Yeah I see a a ton of value in spot, but those characteristics are pretty unpredictable (and different from instances). Terraform aims to be declarative. If my plan states that I have 5 instances running and they disappear some time later, that declarative model breaks.

On the other hand if I think of the **bid** itself as a resource, the declarative mental model stays in tact -- similar to ASGs. I think we can DRY by refactoring so don't be afraid to copy-paste some code to get the mental model right.

Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: Alright, I'll crest it as a separate resource so that we can see what that
would look like, and DRY things up later

File: builtin/providers/aws/resource_aws_instance.go (hunk 1 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: @cbednarski that's an interesting proposition. 

If we don't wait for the bid to resolve, there's no guarantee that the resource will ever actually be provided, which kind of sucks. On the other hand, it would be nice to be able to put the bid in proceed irregardless of if the bid ever resolves.

What do you think about making this another switch? Something like "spot_wait" to toggle if you should wait for the instance to come up or not. I can see use cases for both.

File: builtin/providers/aws/resource_aws_instance.go (hunk 2 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: > On the other hand if I think of the bid itself as a resource

So do you mean that an instance would reference a bid resource? Ie, you would define them separately from one another? If the bid resolves, then the instance is created?

Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: Also, by ASG do you mean Amazon Security Groups?

Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: > It is clearer to see from actually running Terraform, but Terraform creates the EC2 instance before the elastic IP address. Due to the interpolation earlier where the elastic IP requires the ID of the EC2 instance, Terraform is able to infer a dependency, and knows to create the instance first.

Ah, so if I specify a 'spot_bid' parameter in the instance and have it reference the spot bid, it will create that resource first implicitly - that's neat. I'm going to go this route as I agree it's much cleaner.

File: builtin/providers/aws/resource_aws_instance.go (hunk 1 above)
Comment (mitchellh) [inline builtin/providers/aws/resource_aws_instance.go]: It is more Terraform-like to just wait for the bid for now. Terraform waits for things to be fully ready  at a resource level. A spot instance isn't ready until it exists.

Let's get this merged and then we might want to entertain adding another flag to the resource that just tells Terraform to just wait for the bid and move on.

File: builtin/providers/aws/resource_aws_instance.go (hunk 2 above)
Comment (mitchellh) [inline builtin/providers/aws/resource_aws_instance.go]: Hm, so making spot instances a separate resource is really interesting. At the same time, all the fields/logic in `aws_instance` are pretty tricky and I'm not sure how much I want to duplicate those. 

What @cbednarski noticed is a real thing: a spot instance can be pre-empted. But I think anyone using spot instances will know this. Perhaps the best solution would be to remove the fallback for now? Just fail if the spot requests fails?

Comment (catsby) [inline builtin/providers/aws/resource_aws_instance.go]: > Also, by ASG do you mean Amazon Security Groups?

I believe he was speaking about Auto Scaling Groups :smile:

File: builtin/providers/aws/resource_aws_instance.go (hunk 1 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: Cool, I'm working moving spot bids into their own resource right now - i'll make sure that instances wait for the bid to resolve (i had made this a flag)

File: builtin/providers/aws/resource_aws_instance.go (hunk 2 above)
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: > Hm, so making spot instances a separate resource is really interesting. At the same time, all the fields/logic in aws_instance are pretty tricky and I'm not sure how much I want to duplicate those.

This was what scared me at first @mitchellh, but I think we can actually refactor this out pretty easily to be shared by both resources now that I look at it more closely. If we move that logic out to its own function, it can be called easily by either the spot bid OR the on demand instance, and then they can just perform assignments to their request structs as needed.

I haven't done this in my separate-resource rework rough cut (which I'll push as soon as I finish some more testing), but I think it would be a requirement for getting this PR merged.

**Comment by dalehamel** [comment]
closing in favor of spot bids as a separate resource, which is provided by https://github.com/hashicorp/terraform/pull/2109. We can reopen this if we decide to back-peddle.

**Comment by dalehamel** [comment]
As discussed in https://github.com/hashicorp/terraform/pull/2109, this approach is simpler than a separate resource. I've re-opened the PR and I've done some cleanup and fixed most of the FIXME's

Aside from missing tests (there are none, but I've started to point out the ones I expect to have to write), I think this is now ready for a more serious and thorough code review.

Behavior should be:

* If a spot_price is specified, the launch strategy is to request a spot instance **and block waiting for it to resolve**. If it cannot resolve (or times out - note **timeout still not configurable**), then the resource creation will fail.
* Deleting the resource should delete the spot request as well - this is particularly important for persistent spot requests, as the instance can and will come back from the dead if the spot request is left around and the instance is terminated
* "read" should update and store the spot instance status

Usage is like:

```
resource "aws_instance" "test" {
    ami = "ami-50948838"
    instance_type = "m1.small"
    spot_price = "0.05"
    spot_persist = true
    key_name = "dale.hamel"
    tags {
        Name = "HelloWorld"
    }
}
```

@cbednarski @mitchellh @catsby for review

File: builtin/providers/aws/resource_aws_instance.go [hunk 3]
```diff
@@ -749,13 +892,36 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
...
+
+		if err != nil {
+			return fmt.Errorf(
+				"Error cancelling spot reservation (%s): %s",
+				reqID, err)
+		}
+
+		if output != nil {
+			cancelled := output.CancelledSpotInstanceRequests
+			if len(cancelled) != 1 || *cancelled[0].SpotInstanceRequestID != reqID {
```
Comment (dalehamel) [inline builtin/providers/aws/resource_aws_instance.go]: I'm a little wary of this logic myself. It might cause the stack to fail to delete if the spot instance request wasn't fulfilled correctly (I'm not sure how cancel responds to requests that are already cancelled / in some other bad state)

What do you guys think we should do here?

**Comment by dalehamel** [comment]
FYI i tried writing some tests... but it's very difficult given that the tests I saw assume an instance has already been launched, and there's no way to access the resource scheme from the tests as far as I can tell.

I also read that writing acceptance tests are not required to accept the PR, so maybe the manual testing I have done is sufficient?

If that's the case, I guess we're down to just nitpicking the code?

**Comment by dalehamel** [comment]
Anything needed from me to help get this reviewed and merged?

**Comment by mitchellh** [comment]
Nope we'll do so today :)

**Comment by dalehamel** [comment]
> Nope we'll do so today :)

\o/

**Comment by ketzacoatl** [comment]
super exciting, great work @dalehamel!!

**Comment by dalehamel** [comment]
Thanks!

I do feel weird not writing any tests for this though.

I would love to pair with someone more familiar with the testing framework
and the project to submit some tests to make sure that all cases and states
are covered

**Comment by dalehamel** [comment]
bump - any movement on this?

**Comment by dalehamel** [comment]
sorry to keep bumping guys, we'd like to ship this internally and don't want to have to fork terraform.

Is there anything further I can do to facilitate the review and help get this :ship: :dash: ?

cc @mitchellh

**Comment by phinze** [comment]
Hi @dalehamel - sorry for the radio silence here - bump received! We'll get this moving today one way or another. :+1:

**Comment by dalehamel** [comment]
:sparkles: :heart: :sparkles:

**Comment by catsby** [comment]
Hey @dalehamel I'm sorry for the silence here, it's my fault. I'm re-reviewing this and your other PR just to make sure I grok things. 

In the midsts of researching all this (I admit spot instances are new-ish to me), I found that you can use an Auto Scaling Group to launch spot instances:

- [Launching Spot Instances in Your Auto Scaling Group](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/US-SpotInstances.html)

Terraform has support for the `spot_price` field in Launch Configurations used by ASGs (sadly undocumented, but [it's there][1]). On the surface, to me this seems a more viable route for maintaining spot instances. 

You seem to be well versed here, can you help me understand if AGSs+LC with spot prices is not a good route, or at least the specifics why they are not a viable option for your use case? 

Thanks! 

[1]: https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_launch_configuration.go#L90-L94

**Comment by dalehamel** [comment]
> You seem to be well versed here, can you help me understand if AGSs+LC with spot prices is not a good route, or at least the specifics why they are not a viable option for your use case?

@catsby yeah, i was already aware of the ASG support and it's great to see that exists.

Unfortunately, I do think that there's a very valid separate use case for supporting standalone spot instances that is not supported by the autoscaling groups.

The biggest one I can think of is simplicity, if you want to have a very simple stack you shouldn't have to add an autoscaling group with one spot instance - that's just a hack. It's an additional level of indirection.

For longer running instances that just happen to be spot, where you only ever want one of them, creating an autoscaling group is a silly complication, and extra layer of indirection.

**Comment by ketzacoatl** [comment]
@catsby, as a heavy user of ASG and AWS instances in Terraform, but also needing to use spot pricing.. I agree with @dalehamel on this.. it makes the most sense for _both_ resource types (aws instance and asg) to support spot pricing.

Thank you for your work to review this PR!

**Comment by dalehamel** [comment]
@ketzacoatl just to be sure you are aware, you can already use spot pricing with ASGs in terraform. (I think that was linked to in the issue i referenced above).

Can you clarify that you also need individual instances to have spot prices?

**Comment by ketzacoatl** [comment]
Yes @dalehamel, I can confirm wanting spot prices for individual instances. I think spot pricing should be supported on both resources. I was happy to find out about the existing support for ASG, and I'm excited about (and grateful for) your work here. thanks!

**Comment by phinze** [comment]
Hi @dalehamel - just pushed #2263 with a take on this that incorporates your code along with the latest from the core team's discussion about how best to model this.

Since we've again shifted to "additional resource is better" - my idea there was to pick up the coding work instead of asking you to yet again change tacks.

Have a look and let me know what you think!

**Comment by phinze** [comment]
Closing this since we finished it up in #2263 - thanks again for all your hard work on this @dalehamel!

**Comment by dalehamel** [comment]
:clap:
